src/flint/flint_base/flint_base.pyx

```diff
@@ -114,16 +119,117 @@ cdef class flint_poly(flint_elem):
 
 
+cdef class flint_mpoly_context(flint_elem):
+    """
+    Base class for multivariate ring contexts
+    """
+
+    _ctx_cache = None
+
+    def __cinit__(self):
+        self._init = False
+
+    def __init__(self, int nvars, names):
+        if nvars < 0:
+            raise ValueError("cannot have a negative amount of variables")
+        elif len(names) != nvars:
+            raise ValueError("number of variables must match lens of variable names")
+        self.py_names = tuple(name.encode("ascii") if not isinstance(name, bytes) else name for name in names)
+        self.c_names = <const char**> libc.stdlib.malloc(nvars * sizeof(const char *))
+        self._init = True
+        for i in range(nvars):
+            self.c_names[i] = self.py_names[i]
+
+    def __dealloc__(self):
+        if self._init:
+            libc.stdlib.free(self.c_names)
+        self._init = False
+
+    def __str__(self):
+        return self.__repr__()
+
+    def __repr__(self):
+        return f"{self.__class__.__name__}({self.nvars()}, '{self.ordering()}', {self.names()})"
+
+    def name(self, long i):
+        if not 0 <= i < len(self.py_names):
+            raise IndexError("variable name index out of range")
+        return self.py_names[i].decode('utf-8')
+
+    def names(self):
+        return tuple(name.decode('utf-8') for name in self.py_names)
+
+    def gens(self):
+        return tuple(self.gen(i) for i in range(self.nvars()))
+
+    @staticmethod
+    def create_cache_key(slong nvars, ordering: str, names: str):
+        nametup = tuple(name.strip() for name in names.split(','))
+        if len(nametup) != nvars:
+            if len(nametup) != 1:
+                raise ValueError("Number of variables does not equal number of names")
+            nametup = tuple(nametup[0] + str(i) for i in range(nvars))
+        return nvars, ordering, nametup
+
+    @classmethod
+    def get_context(cls, slong nvars=1, ordering: str = "lex", names: Optional[str] = "x", nametup: Optional[tuple] = None):
+        if nvars < 0:
+            raise ValueError("cannot have a negative amount of variables")
+
+        if nametup is None and names is not None:
+            key = cls.create_cache_key(nvars, ordering, names)
+        elif len(nametup) != nvars:
+            raise ValueError("Number of variables does not equal number of names")
+        else:
+            key = (nvars, ordering, nametup)
+
+        ctx = cls._ctx_cache.get(key)
+        if ctx is None:
+            ctx = cls(*key)
+            cls._ctx_cache[key] = ctx
+        return ctx
+
+    @classmethod
+    def from_context(cls, ctx: flint_mpoly_context):
+        return cls.get_context(
+            nvars=ctx.nvars(),
+            ordering=ctx.ordering(),
+            names=None,
+            nametup=tuple(str(s, 'utf-8') for s in ctx.py_names)
+        )
+
+    @classmethod
+    def joint_context(cls, *ctxs):
+        vars = set()
+        for ctx in ctxs:
```

Comment:
Is this used anywhere?

I would remove this...

Reply:
> Is this used anywhere?

Not any more